pass_args_to_sdf.md

```diff
@@ -0,0 +1,201 @@
+
+In a launch/world file, the user can construct a model using model composition and include components of the model using the `<include>` tag. If a user wants to modify parameters of the included model, then the user will use the custom element `<gazebo:params>` that will contain the adjusted elements. In the custom element `<gazebo:params>`, `gazebo` is the namespace prefix and `params` is the custom element. This approach was chosen so that downstream users can choose to ignore the namespaced custom element and its contents. The included model (not the constructed model in the launch/world file) will be referred to as the original model/file. In the launch/world file and under `<include>`, the elements listed under `<gazebo:params>` will indicate elements from the original model and will specify new values and/or elements to be updated, added, and/or removed.
+
+To specify an element from the original file, the tag (e.g., `<link>`, `<sensor>`, `<visual>`) needs to be provided as well as the name attribute but the name is the name of all the parent elements including its own name separated by double colon (`::`) syntax (e.g., `name="chassis::camera::cam"` where "chassis" is the name of the `<link>`, "camera" is the name of the `<sensor>`, and "cam" is the name of the `<camera>`). This is how the correct element will be identified and will be called the element identifier. Although it is possible to identify the element without requiring the user to specify the tag, enforcing the user to provide the tag allows less ambiguity to the user as well as provides a means of error checking. Then the user will need to specify a corresponding action with that element to dictate the alteration they would like to make using a custom attribute `action`. This attribute takes in a string and the available actions at the specified element are:
+
+* `modify`: indicates that the values in the original model are to be modified to the new listed values
+* `replace`: replaces the elements and/or values from the original model to the new provided elements and/or values
+* `add`: adds new elements to the original model
+* `remove`: removes (or disables) the elements from the original model
```

Comment:
seeing these actions listed out reminds me of the functionality of the [sdf::Converter class](https://github.com/osrf/sdformat/blob/sdformat10_10.0.0/src/Converter.hh#L67-L104) (see also [sdf/1.6/1_5.convert](https://github.com/osrf/sdformat/blob/sdformat10_10.0.0/sdf/1.6/1_5.convert) and [sdf/1.7/1_6.convert](https://github.com/osrf/sdformat/blob/sdformat10_10.0.0/sdf/1.7/1_6.convert)). it's currently internal to libsdformat, so you couldn't call it as an API, but the code itself could be useful to refer to or to fork into your own implementation

Reply:
The same could be done using the [Element class](http://osrf-distributions.s3.amazonaws.com/sdformat/api/9.0.0/classsdf_1_1v9_1_1Element.html) correct (e.g., `AddElement`, `RemoveChild`, `Set`)?

I guess it comes down to if we want to implement this in ign-gazebo or in libsdformat? If we implement this in ign-gazebo, then libsdformat's `Element` class could be used. If in libsdformat, then some new function (e.g., `LoadNewArgs`) in the [Root class](https://github.com/osrf/sdformat/blob/sdformat10_10.0.0/include/sdf/Root.hh) could use `sdf::Converter`.

Reply:
sure, the existing `Element` APIs could be used; I just meant to say that the `Converter` class is already doing something similar, so it would be a shame to have to maintain two different approaches to doing the same type of xml transformation. If it's feasible to share code, I would modify the `sdf::Converter` class to support that, but they can also remain separate if it doesn't make sense to share code